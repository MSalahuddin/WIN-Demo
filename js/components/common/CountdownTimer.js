import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Text from './Text';
import { formatTimeString } from '../../utils';

const CountdownTimer = ({ time, onTimerEnd, onFiveSecondsLeft, ...rest }) => {
  const [remainingTime, setRemainingTime] = useState(time);
  useEffect(() => {
    const countdownInterval = setInterval(() => {
      if (remainingTime > 0) {
        setRemainingTime(remainingTime - 1);
      }
    }, 1000);

    if (remainingTime === 5) {
      onFiveSecondsLeft();
    }

    if (remainingTime === 0) {
      onTimerEnd();
      clearInterval(countdownInterval);
    }
    return () => {
      clearInterval(countdownInterval);
    };
  }, [onFiveSecondsLeft, onTimerEnd, remainingTime]);

  const formatString = formatTimeString(remainingTime);
  return <Text allowFontScaling={false} {...rest}>{formatString}</Text>;
};

CountdownTimer.propTypes = {
  time: PropTypes.number.isRequired,
  onTimerEnd: PropTypes.func,
  onFiveSecondsLeft: PropTypes.func
};

CountdownTimer.defaultProps = {
  onTimerEnd: () => {},
  onFiveSecondsLeft: () => {}
};

export default CountdownTimer;
