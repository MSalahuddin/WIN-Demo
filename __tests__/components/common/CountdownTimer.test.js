import React from 'react';
import { render } from 'react-native-testing-library';

import CountdownTimer from '../../../js/components/common/CountdownTimer';

describe('Component CountdownTimer', () => {
  const onTimerEnd = jest.fn();
  const onFiveSecondsLeft = jest.fn();

  test('renders correctly with 6 seconds', () => {
    jest.useFakeTimers();
    const output = render(<CountdownTimer time={6} onTimerEnd={onTimerEnd} onFiveSecondsLeft={onFiveSecondsLeft} />);
    expect(output).toMatchSnapshot();

    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(onFiveSecondsLeft).not.toBeCalled();
    jest.runAllTimers();
    expect(clearInterval).toBeCalled();
    expect(onFiveSecondsLeft).toBeCalled();
    expect(onTimerEnd).toBeCalled();
  });
});
