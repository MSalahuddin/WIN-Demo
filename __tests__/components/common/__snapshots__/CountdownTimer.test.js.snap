// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component CountdownTimer renders correctly with 6 seconds 1`] = `
<Text
  alignCenter={false}
  allowFontScaling={false}
  color="#000000"
  fontFamily="MEDIUM"
  isUnderlined={false}
  isUppercase={false}
  maxFontSizeMultiplier={1.3}
  size="NORMAL"
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "Calibre-Medium",
        "fontSize": 30,
        "justifyContent": "center",
        "textAlign": "left",
        "textDecorationColor": "#000000",
        "textDecorationLine": "none",
        "textDecorationStyle": "solid",
        "textTransform": "none",
      },
    ]
  }
>
  00:06
</Text>
`;
